File: macos-native/Sources/Core/TtsService.swift
import AVFoundation

/// Fire-and-forget text-to-speech — neural TTS via edge-tts CLI, with
/// AVSpeechSynthesizer as fallback when edge-tts is unavailable.
final class TtsService: NSObject, AVSpeechSynthesizerDelegate {
    static let shared = TtsService()

    private let synthesizer = AVSpeechSynthesizer()

    /// Active afplay process for single-slot cancellation.
    private var activeProcess: Process?

    private override init() {
        super.init()
        synthesizer.delegate = self
    }

    // MARK: - Public

    func speak(_ text: String) {
        let config = ConfigStore.shared
        guard config.ttsEnabled else { return }
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        // Cancel any previous speech
        stop()

        Task.detached { [weak self] in
            let tmpPath = NSTemporaryDirectory() + "vibetotext_tts.mp3"
            let voice = config.ttsVoice ?? "en-GB-RyanNeural"
            let rate = config.ttsEdgeRate ?? "+12%"
            let pitch = config.ttsEdgePitch ?? "+1Hz"

            // 1. Generate mp3 via edge-tts CLI
            let genProcess = Process()
            genProcess.executableURL = URL(fileURLWithPath: "/usr/bin/env")
            genProcess.arguments = [
                "edge-tts",
                "--voice", voice,
                "--rate", rate,
                "--pitch", pitch,
                "--text", text,
                "--write-media", tmpPath,
            ]
            genProcess.standardOutput = FileHandle.nullDevice
            genProcess.standardError = FileHandle.nullDevice

            do {
                try genProcess.run()
                genProcess.waitUntilExit()

                guard genProcess.terminationStatus == 0 else {
                    print("[TTS] edge-tts exited with status \(genProcess.terminationStatus), falling back")
                    self?.speakFallback(text)
                    return
                }

                // 2. Play with afplay
                let playProcess = Process()
                playProcess.executableURL = URL(fileURLWithPath: "/usr/bin/afplay")
                playProcess.arguments = [tmpPath]
                try playProcess.run()
                self?.activeProcess = playProcess
            } catch {
                print("[TTS] edge-tts failed: \(error), falling back")
                self?.speakFallback(text)
            }
        }
    }

    func stop() {
        // Kill active afplay process
        if let proc = activeProcess, proc.isRunning {
            proc.terminate()
        }
        activeProcess = nil

        // Also stop AVSpeechSynthesizer fallback
        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }
    }

    // MARK: - Fallback (AVSpeechSynthesizer)

    private func speakFallback(_ text: String) {
        let config = ConfigStore.shared

        // Cancel any previous fallback speech
        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }

        let utterance = AVSpeechUtterance(string: text)

        // Convert WPM (default 200) to AVSpeechUtterance rate (0.0 - 1.0)
        let wpm = Float(config.ttsRate)
        utterance.rate = max(AVSpeechUtteranceMinimumSpeechRate,
                            min(AVSpeechUtteranceMaximumSpeechRate,
                                (wpm / 200.0) * AVSpeechUtteranceDefaultSpeechRate))

        // Volume: 0-100 -> 0.0-1.0
        utterance.volume = Float(config.ttsVolume) / 100.0

        let voiceId = (config.ttsVoice?.isEmpty == false) ? config.ttsVoice : nil
        if let id = voiceId {
            utterance.voice = AVSpeechSynthesisVoice(identifier: id)
                ?? AVSpeechSynthesisVoice(language: id)
        } else {
            // Default: Daniel (British male) for Jarvis feel
            utterance.voice = AVSpeechSynthesisVoice(identifier: "com.apple.voice.compact.en-GB.Daniel")
                ?? AVSpeechSynthesisVoice(language: "en-GB")
        }

        synthesizer.speak(utterance)
    }

    // MARK: - Status messages

    static func generateStatusMessage(mode: String, text: String, output: String, fileCount: Int = 0) -> String {
        switch mode {
        case "greppy":
            return fileCount == 1 ? "Located one file, sir" : "Located \(fileCount) files, sir"
        case "cleanup":
            let n = output.components(separatedBy: "\n\n")
                .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
                .count
            return "All tidied up. \(n) paragraphs ready"
        case "plan":
            let n = output.components(separatedBy: "\n")
                .filter { line in
                    let trimmed = line.trimmingCharacters(in: .whitespaces)
                    return trimmed.range(of: #"^(\d+[\.\):]|[-*])\s"#, options: .regularExpression) != nil
                }
                .count
            return "Plan's ready. \(n) steps laid out"
        case "feedback":
            return "Feedback spoken, sir"
        default:
            let n = text.split(separator: " ").count
            return "Got it. \(n) words captured"
        }
    }
}

File: windows-native/src/VibeToText/Core/TtsService.cs
using System.Diagnostics;
using System.IO;
using System.Speech.Synthesis;
using VibeToText.Data;

namespace VibeToText.Core;

/// <summary>
/// Fire-and-forget text-to-speech for status reports.
/// Primary: edge-tts (neural voice) + ffplay for playback.
/// Fallback: System.Speech.Synthesis (built-in SAPI).
/// </summary>
public class TtsService : IDisposable
{
    private SpeechSynthesizer? _synth;
    private readonly ConfigStore _config;
    private readonly object _lock = new();
    private Process? _ffplayProcess;
    private static readonly string TempMp3Path = Path.Combine(Path.GetTempPath(), "vibetotext_tts.mp3");

    // edge-tts defaults
    private const string DefaultVoice = "en-GB-RyanNeural";

    public TtsService(ConfigStore config)
    {
        _config = config;
        try
        {
            _synth = new SpeechSynthesizer();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[TTS] Failed to initialize SAPI fallback: {ex.Message}");
        }
    }

    /// <summary>
    /// Speak text using edge-tts (neural) with SAPI fallback. Fire-and-forget.
    /// </summary>
    public void Speak(string text)
    {
        if (!_config.TtsEnabled || string.IsNullOrWhiteSpace(text))
            return;

        // Cancel any previous speech
        StopPlayback();

        // Run everything in background so we don't block
        Task.Run(() =>
        {
            try
            {
                SpeakEdgeTts(text);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[TTS] edge-tts failed, falling back to SAPI: {ex.Message}");
                try
                {
                    SpeakSapiFallback(text);
                }
                catch (Exception ex2)
                {
                    Console.WriteLine($"[TTS] SAPI fallback also failed: {ex2.Message}");
                }
            }
        });
    }

    /// <summary>
    /// Stop current playback (kill ffplay or cancel SAPI).
    /// </summary>
    public void StopPlayback()
    {
        lock (_lock)
        {
            // Kill ffplay if running
            if (_ffplayProcess != null)
            {
                try
                {
                    if (!_ffplayProcess.HasExited)
                        _ffplayProcess.Kill();
                    _ffplayProcess.Dispose();
                }
                catch { }
                _ffplayProcess = null;
            }

            // Cancel SAPI if running
            try { _synth?.SpeakAsyncCancelAll(); }
            catch { }
        }
    }

    private void SpeakEdgeTts(string text)
    {
        var voice = _config.TtsVoice;
        if (string.IsNullOrEmpty(voice))
            voice = DefaultVoice;
        var rate = _config.TtsEdgeRate;
        var pitch = _config.TtsEdgePitch;

        // Step 1: Generate mp3 with edge-tts CLI
        var edgeProcess = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "edge-tts",
                Arguments = $"--voice \"{voice}\" --rate \"{rate}\" --pitch \"{pitch}\" --text \"{EscapeArg(text)}\" --write-media \"{TempMp3Path}\"",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            }
        };

        edgeProcess.Start();
        // Consume stdout/stderr to prevent deadlock
        edgeProcess.StandardOutput.ReadToEnd();
        edgeProcess.StandardError.ReadToEnd();

        if (!edgeProcess.WaitForExit(15_000)) // 15 second timeout
        {
            try { edgeProcess.Kill(); } catch { }
            throw new TimeoutException("edge-tts timed out after 15 seconds");
        }

        if (edgeProcess.ExitCode != 0)
            throw new InvalidOperationException($"edge-tts exited with code {edgeProcess.ExitCode}");

        if (!File.Exists(TempMp3Path) || new FileInfo(TempMp3Path).Length == 0)
            throw new FileNotFoundException("edge-tts did not produce an output file");

        // Step 2: Play with ffplay
        var ffplay = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "ffplay",
                Arguments = $"-nodisp -autoexit -loglevel quiet \"{TempMp3Path}\"",
                UseShellExecute = false,
                CreateNoWindow = true,
            }
        };

        lock (_lock)
        {
            ffplay.Start();
            _ffplayProcess = ffplay;
        }

        // Wait for playback to complete (don't leave zombie processes)
        ffplay.WaitForExit();
    }

    private void SpeakSapiFallback(string text)
    {
        if (_synth == null)
            throw new InvalidOperationException("SAPI SpeechSynthesizer not available");

        // Apply settings
        // SAPI rate: -10 to 10, WPM 185 ~ -1 (slightly slower for Jarvis feel)
        _synth.Rate = Math.Clamp((_config.TtsRate - 200) / 20, -10, 10);
        _synth.Volume = Math.Clamp(_config.TtsVolume, 0, 100);

        var voice = _config.TtsVoice;
        if (string.IsNullOrEmpty(voice))
            voice = "Microsoft David Desktop"; // Deep male - Jarvis vibe
        try { _synth.SelectVoice(voice); }
        catch { /* voice not found, use default */ }

        _synth.SpeakAsync(text);
    }

    /// <summary>Escape double quotes in text for CLI argument.</summary>
    private static string EscapeArg(string text)
    {
        return text
            .Replace("\\", "\\\\")
            .Replace("\"", "\\\"")
            .Replace("\r", " ")
            .Replace("\n", " ");
    }

    public static string GenerateStatusMessage(RecordingMode mode, string text, string output)
    {
        return mode switch
        {
            RecordingMode.Greppy => "Files located, sir",
            RecordingMode.Cleanup => $"All tidied up. {CountParagraphs(output)} paragraphs ready",
            RecordingMode.Plan => $"Plan's ready. {CountSteps(output)} steps laid out",
            RecordingMode.Feedback => "Feedback spoken, sir",
            _ => $"Got it. {CountWords(text)} words captured",
        };
    }

    private static int CountWords(string text) =>
        text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries).Length;

    private static int CountParagraphs(string text) =>
        text.Split("\n\n", StringSplitOptions.RemoveEmptyEntries)
            .Count(p => !string.IsNullOrWhiteSpace(p));

    private static int CountSteps(string text) =>
        text.Split('\n')
            .Count(line =>
            {
                var trimmed = line.TrimStart();
                return System.Text.RegularExpressions.Regex.IsMatch(trimmed, @"^(\d+[\.\):]|[-*])\s");
            });

    public void Dispose()
    {
        StopPlayback();
        try
        {
            _synth?.Dispose();
        }
        catch { }
        _synth = null;

        // Clean up temp file
        try
        {
            if (File.Exists(TempMp3Path))
                File.Delete(TempMp3Path);
        }
        catch { }
    }
}
